<img src=http://multimodal.mipa.ugm.ac.id/wp-content/uploads/sites/371/2020/02/Banner-E-Course-Isna-2-01.jpg>

<img src="https://image.freepik.com/free-photo/chicken-wings-barbecue-sweetly-sour-sauce-picnic-summer-menu-tasty-food-top-view-flat-lay_2829-6470.jpg" width=800px height=200px>

Sebelum mempersiapkan data, terlebih dahulu kita harus menyiapkan library/package terkait yang kita butuhkan.. 

In [1]:
import pandas as pd # kita akan import library pandas untuk menggunakan dataframe pandas
import numpy as np # import library numpy
import re # import library regular expression (regex)

Selanjutnya kita siapkan dataset. Pada contoh ini kita akan menggunakan dataset resep yang diambil dari kelas Cognitif AI. https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DS0103EN/labs/data/recipes.csv

In [2]:
recipes = pd.read_csv("recipes.csv")

Selanjutnya, kita cek sample data nya..

In [3]:
recipes.head()

country almond angelica anise anise_seed apple apple_brandy apricot  \
0  Vietnamese     No       No    No         No    No           No      No   
1  Vietnamese     No       No    No         No    No           No      No   
2  Vietnamese     No       No    No         No    No           No      No   
3  Vietnamese     No       No    No         No    No           No      No   
4  Vietnamese     No       No    No         No    No           No      No   

  armagnac artemisia  ... whiskey white_bread white_wine  \
0       No        No  ...      No          No         No   
1       No        No  ...      No          No         No   
2       No        No  ...      No          No         No   
3       No        No  ...      No          No         No   
4       No        No  ...      No          No         No   

  whole_grain_wheat_flour wine wood yam yeast yogurt zucchini  
0                      No   No   No  No    No     No       No  
1                      No   No   No  No    No     No       No  
2                      No   No   No  No    No     No       No  
3                      No   No   No  No    No     No       No  
4                      No   No   No  No    No     No       No  

[5 rows x 384 columns]

Dari hasil eksekusi kode, tampak bahwa ada 5 baris sample data dengan 384 kolom / fitur. Ada 384 bahan yang digunakan dalam dataset ini, dimana penggunaan bahan dalam masakan direpsentasikan dengan kategori 'Yes' atau 'No'.

Kita juga bisa melihat dimensi dari dataset kita. 

In [4]:
recipes.shape

(57691, 384)

Dari hasil eksekusi dataset ini memiliki 57691 baris data dan 384 kolom. 

Nah berikutnya, kita mau cek apakah data kita memiliki bahan-bahan tertentu. Misal, mau cek ada ikan (fish) nggak si? 
Beras yang ada apa saja ya? dst.. 

In [5]:
ingredients = list(recipes.columns.values)

print([match.group(0) for ingredient in ingredients for match in \
       [(re.compile(".*(rice).*")).search(ingredient)] if match])
print([match.group(0) for ingredient in ingredients for match in \
       [(re.compile(".*(fish).*")).search(ingredient)] if match])

['brown_rice', 'licorice', 'rice']
['catfish', 'fish', 'shellfish', 'smoked_fish']


Oke sekaran saatnya menghitung komposisi dataset.. ada berapa resep sih per negara? 

In [6]:
recipes["country"].value_counts() # komposisi

American                 40150
Mexico                    1754
Italian                   1715
Italy                     1461
Asian                     1176
French                     996
east_asian                 951
Canada                     774
korean                     767
Mexican                    622
western                    450
Southern_SoulFood          346
India                      324
Jewish                     320
Spanish_Portuguese         291
Mediterranean              289
UK-and-Ireland             282
Indian                     274
France                     268
MiddleEastern              248
Central_SouthAmerican      241
Germany                    237
Eastern-Europe             235
Chinese                    226
Greek                      225
English_Scottish           204
Caribbean                  183
Thai                       164
Scandinavia                158
Cajun_Creole               146
                         ...  
Scandinavian                92
chinese 

***Langkah data cleaning pertama: Ubah kolom negara menjadi masakan. / cuisine***

In [7]:
recipes.head()

country almond angelica anise anise_seed apple apple_brandy apricot  \
0  Vietnamese     No       No    No         No    No           No      No   
1  Vietnamese     No       No    No         No    No           No      No   
2  Vietnamese     No       No    No         No    No           No      No   
3  Vietnamese     No       No    No         No    No           No      No   
4  Vietnamese     No       No    No         No    No           No      No   

  armagnac artemisia  ... whiskey white_bread white_wine  \
0       No        No  ...      No          No         No   
1       No        No  ...      No          No         No   
2       No        No  ...      No          No         No   
3       No        No  ...      No          No         No   
4       No        No  ...      No          No         No   

  whole_grain_wheat_flour wine wood yam yeast yogurt zucchini  
0                      No   No   No  No    No     No       No  
1                      No   No   No  No    No     No       No  
2                      No   No   No  No    No     No       No  
3                      No   No   No  No    No     No       No  
4                      No   No   No  No    No     No       No  

[5 rows x 384 columns]

In [8]:
column_names = recipes.columns.values
column_names[0] = "cuisine"
recipes.columns = column_names

recipes

cuisine almond angelica anise anise_seed apple apple_brandy apricot  \
0      Vietnamese     No       No    No         No    No           No      No   
1      Vietnamese     No       No    No         No    No           No      No   
2      Vietnamese     No       No    No         No    No           No      No   
3      Vietnamese     No       No    No         No    No           No      No   
4      Vietnamese     No       No    No         No    No           No      No   
5      Vietnamese     No       No    No         No    No           No      No   
6      Vietnamese     No       No    No         No    No           No      No   
7      Vietnamese     No       No    No         No    No           No      No   
8      Vietnamese     No       No    No         No    No           No      No   
9      Vietnamese     No       No    No         No    No           No      No   
10     Vietnamese     No       No    No         No    No           No      No   
11     Vietnamese     No       No    No         No    No           No      No   
12     Vietnamese     No       No    No         No    No           No      No   
13     Vietnamese     No       No    No         No    No           No      No   
14     Vietnamese     No       No    No         No    No           No      No   
15     Vietnamese     No       No    No         No    No           No      No   
16     Vietnamese     No       No    No         No    No           No      No   
17     Vietnamese     No       No    No         No    No           No      No   
18     Vietnamese     No       No    No         No    No           No      No   
19     Vietnamese     No       No    No         No    No           No      No   
20     Vietnamese     No       No    No         No    No           No      No   
21     Vietnamese     No       No    No         No    No           No      No   
22     Vietnamese     No       No    No         No    No           No      No   
23     Vietnamese     No       No    No         No    No           No      No   
24     Vietnamese     No       No    No         No    No           No      No   
25     Vietnamese     No       No    No         No    No           No      No   
26     Vietnamese     No       No    No         No    No           No      No   
27     Vietnamese     No       No    No         No    No           No      No   
28     Vietnamese     No       No    No         No    No           No      No   
29     Vietnamese     No       No    No         No    No           No      No   
...           ...    ...      ...   ...        ...   ...          ...     ...   
57661       Japan     No       No    No         No    No           No      No   
57662       Japan     No       No    No         No    No           No      No   
57663       Japan     No       No    No         No    No           No      No   
57664       Japan     No       No    No         No    No           No      No   
57665       Japan     No       No    No         No    No           No      No   
57666       Japan     No       No    No         No    No           No      No   
57667       Japan     No       No    No         No    No           No      No   
57668       Japan     No       No    No         No    No           No      No   
57669       Japan     No       No    No         No    No           No      No   
57670       Japan     No       No    No         No    No           No      No   
57671       Japan     No       No    No         No    No           No      No   
57672       Japan     No       No    No         No    No           No      No   
57673       Japan     No       No    No         No    No           No      No   
57674       Japan     No       No    No         No    No           No      No   
57675       Japan     No       No    No         No    No           No      No   
57676       Japan     No       No    No         No    No           No      No   
57677       Japan     No       No    No         No    No           No      No   
57678       Japan     No       No    No  

Ubah nama cuisine jadi lowercase

In [9]:
recipes["cuisine"] = recipes["cuisine"].str.lower()

Nah selanjutnya, dari data resep ada beberapa cuisine yang ditulis dengan istilah yang berbeda padahal sama. Misal Indonesia, Indonesian. dst. Ini perlu kita rapihkan menjadi satu nama per cuisine. 

In [10]:
recipes.loc[recipes["cuisine"] == "austria", "cuisine"] = "austrian"
recipes.loc[recipes["cuisine"] == "belgium", "cuisine"] = "belgian"
recipes.loc[recipes["cuisine"] == "china", "cuisine"] = "chinese"
recipes.loc[recipes["cuisine"] == "canada", "cuisine"] = "canadian"
recipes.loc[recipes["cuisine"] == "netherlands", "cuisine"] = "dutch"
recipes.loc[recipes["cuisine"] == "france", "cuisine"] = "french"
recipes.loc[recipes["cuisine"] == "germany", "cuisine"] = "german"
recipes.loc[recipes["cuisine"] == "india", "cuisine"] = "indian"
recipes.loc[recipes["cuisine"] == "indonesia", "cuisine"] = "indonesian"
recipes.loc[recipes["cuisine"] == "iran", "cuisine"] = "iranian"
recipes.loc[recipes["cuisine"] == "italy", "cuisine"] = "italian"
recipes.loc[recipes["cuisine"] == "japan", "cuisine"] = "japanese"
recipes.loc[recipes["cuisine"] == "israel", "cuisine"] = "jewish"
recipes.loc[recipes["cuisine"] == "korea", "cuisine"] = "korean"
recipes.loc[recipes["cuisine"] == "lebanon", "cuisine"] = "lebanese"
recipes.loc[recipes["cuisine"] == "malaysia", "cuisine"] = "malaysian"
recipes.loc[recipes["cuisine"] == "mexico", "cuisine"] = "mexican"
recipes.loc[recipes["cuisine"] == "pakistan", "cuisine"] = "pakistani"
recipes.loc[recipes["cuisine"] == "philippines", "cuisine"] = "philippine"
recipes.loc[recipes["cuisine"] == "scandinavia", "cuisine"] = "scandinavian"
recipes.loc[recipes["cuisine"] == "spain", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "portugal", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "switzerland", "cuisine"] = "swiss"
recipes.loc[recipes["cuisine"] == "thailand", "cuisine"] = "thai"
recipes.loc[recipes["cuisine"] == "turkey", "cuisine"] = "turkish"
recipes.loc[recipes["cuisine"] == "vietnam", "cuisine"] = "vietnamese"
recipes.loc[recipes["cuisine"] == "uk-and-ireland", "cuisine"] = "uk-and-irish"
recipes.loc[recipes["cuisine"] == "irish", "cuisine"] = "uk-and-irish"

Nah berikutnya, kita akan pilih2 resep. Resep yang kita pakai kita short misal, cuisine yang > 50 aja. Oke sekarang kita bikin dulu shortist kandidatnya ya.. cusine dengan jumlah resep >50. 

In [11]:
# Keep resep yes..
jml_resep = recipes["cuisine"].value_counts()
cuisines_indeks = jml_resep > 50

cuisines_to_keep = list(np.array(jml_resep.index.values)[np.array(cuisines_indeks)])

Nah sekarang saatnya mengimplementasikan selestion kita tadi ke dataframe kita ya.. 

In [12]:
baris_sebelum = recipes.shape[0] # jumlah ori
print("Number of rows of original dataframe is {}.".format(baris_sebelum))

recipes = recipes.loc[recipes['cuisine'].isin(cuisines_to_keep)]

baris_sesudah = recipes.shape[0] # setelah diproses
print("Number of rows of processed dataframe is {}.".format(baris_sesudah))

print("{} baris yang dihapus!".format(baris_sebelum - baris_sesudah))

Number of rows of original dataframe is 57691.
Number of rows of processed dataframe is 57403.
288 baris yang dihapus!


Berikutnya, untuk mempermudah perhitungan nantinya, kita ubah kategori (yes, no) menjadi angka 1 dan 0. 

In [13]:
recipes = recipes.replace(to_replace="Yes", value=1)
recipes = recipes.replace(to_replace="No", value=0)

Oke Cekidot hasil cleaning kita dulu..

In [14]:
recipes.head(10)

cuisine  almond  angelica  anise  anise_seed  apple  apple_brandy  \
0  vietnamese       0         0      0           0      0             0   
1  vietnamese       0         0      0           0      0             0   
2  vietnamese       0         0      0           0      0             0   
3  vietnamese       0         0      0           0      0             0   
4  vietnamese       0         0      0           0      0             0   
5  vietnamese       0         0      0           0      0             0   
6  vietnamese       0         0      0           0      0             0   
7  vietnamese       0         0      0           0      0             0   
8  vietnamese       0         0      0           0      0             0   
9  vietnamese       0         0      0           0      0             0   

   apricot  armagnac  artemisia  ...  whiskey  white_bread  white_wine  \
0        0         0          0  ...        0            0           0   
1        0         0          0  ...        0            0           0   
2        0         0          0  ...        0            0           0   
3        0         0          0  ...        0            0           0   
4        0         0          0  ...        0            0           0   
5        0         0          0  ...        0            0           0   
6        0         0          0  ...        0            0           0   
7        0         0          0  ...        0            0           0   
8        0         0          0  ...        0            0           0   
9        0         0          0  ...        0            0           0   

   whole_grain_wheat_flour  wine  wood  yam  yeast  yogurt  zucchini  
0                        0     0     0    0      0       0         0  
1                        0     0     0    0      0       0         0  
2                        0     0     0    0      0       0         0  
3                        0     0     0    0      0       0         0  
4                        0     0     0    0      0       0         0  
5                        0     0     0    0      0       0         0  
6                        0     0     0    0      0       0         0  
7                        0     0     0    0      0       0         0  
8                        0     0     0    0      0       0         0  
9                        0     0     0    0      0       0         0  

[10 rows x 384 columns]

In [20]:
recipes.to_csv('clean_recipes.csv',index=False)

Sekarang kita coba cek dari resep2 ini mana yang menggunakan rice dan wasabi ya.. 

In [14]:
check_recipes = recipes.loc[
    (recipes["rice"] == 1) &
    (recipes["wasabi"] == 1) &
    (recipes["seaweed"] == 1) &
    (recipes["soy_sauce"] == 1)
]

check_recipes

cuisine  almond  angelica  anise  anise_seed  apple  apple_brandy  \
11306    japanese       0         0      0           0      0             0   
11321    japanese       0         0      0           0      0             0   
11361    japanese       0         0      0           0      0             0   
12171       asian       0         0      0           0      0             0   
12385       asian       0         0      0           0      0             0   
13010       asian       0         0      0           0      0             0   
13159       asian       0         0      0           0      0             0   
13513    japanese       0         0      0           0      0             0   
13586    japanese       0         0      0           0      0             0   
13625  east_asian       0         0      0           0      0             0   
14495  east_asian       0         0      0           0      0             0   

       apricot  armagnac  artemisia  ...  whiskey  white_bread  white_wine  \
11306        0         0          0  ...        0            0           0   
11321        0         0          0  ...        0            0           0   
11361        0         0          0  ...        0            0           0   
12171        0         0          0  ...        0            0           0   
12385        0         0          0  ...        0            0           0   
13010        0         0          0  ...        0            0           0   
13159        0         0          0  ...        0            0           0   
13513        0         0          0  ...        0            0           0   
13586        0         0          0  ...        0            0           0   
13625        0         0          0  ...        0            0           0   
14495        0         0          0  ...        0            0           0   

       whole_grain_wheat_flour  wine  wood  yam  yeast  yogurt  zucchini  
11306                        0     0     0    0      0       0         0  
11321                        0     1     0    0      0       0         0  
11361                        0     0     0    0      0       0         0  
12171                        0     1     0    0      0       0         0  
12385                        0     0     0    0      0       0         0  
13010                        0     0     0    0      0       0         0  
13159                        0     0     0    0      0       0         0  
13513                        0     0     0    0      0       0         0  
13586                        0     0     0    0      0       0         0  
13625                        0     0     0    0      0       0         0  
14495                        0     0     0    0      0       0         0  

[11 rows x 384 columns]

Sekarang kita hitung jumlah bahannya: 

In [15]:
# jumlah per kolom
bhn = recipes.iloc[:, 1:].sum(axis=0)

In [21]:
# setiap kolom dibuat dalam series pandas
bahan = pd.Series(bhn.index.values, index = np.arange(len(bhn)))
jumlah = pd.Series(list(bhn), index = np.arange(len(bhn)))

# buat dataframe
bhn_df = pd.DataFrame(dict(bahan = bahan, jumlah = jumlah))
bhn_df = bhn_df[["bahan", "jumlah"]]
print(bhn_df.to_string())

                       bahan  jumlah
0                     almond    2306
1                   angelica       1
2                      anise     223
3                 anise_seed      87
4                      apple    2422
5               apple_brandy      37
6                    apricot     620
7                   armagnac      11
8                  artemisia      13
9                  artichoke     391
10                 asparagus     460
11                   avocado     660
12                     bacon    2169
13              baked_potato       9
14                      balm       3
15                    banana     989
16                    barley     266
17             bartlett_pear      23
18                     basil    3842
19                       bay    1463
20                      bean    1992
21                     beech       1
22                      beef    4902
23                beef_broth     845
24                beef_liver      10
25                      beer     307
2

Berikutnya kita cek bahan yang paling banyak digunakan ya.. 

In [22]:
bhn_df.sort_values(["jumlah"], ascending=False, inplace=True)
bhn_df.reset_index(inplace=True, drop=True)

print(bhn_df)

                bahan  jumlah
0                 egg   21025
1               wheat   20781
2              butter   20719
3               onion   18080
4              garlic   17353
5                milk   12870
6       vegetable_oil   11105
7               cream   10171
8              tomato    9920
9           olive_oil    9876
10       black_pepper    9828
11             pepper    9230
12            vanilla    9010
13            cayenne    8254
14            vinegar    8060
15      cane_molasses    7741
16        bell_pepper    5979
17           cinnamon    5594
18            parsley    5552
19            chicken    5436
20        lemon_juice    5065
21               beef    4902
22               corn    4828
23              cocoa    4799
24           scallion    4782
25              bread    4571
26             ginger    4358
27            mustard    4119
28               rice    3857
29              basil    3842
..                ...     ...
353              balm       3
354       

Nah kalau kita perhatikan kebanyakan dari data kita adalah resep amerika.. ini akan jadi bias..
Oke kita cek dulu..

In [66]:
cuisines = recipes.groupby("cuisine").mean()
cuisines.head()

almond  angelica     anise  anise_seed     apple  \
cuisine                                                            
african       0.156522  0.000000  0.000000    0.000000  0.034783   
american      0.040598  0.000025  0.003014    0.000573  0.052055   
asian         0.007544  0.000000  0.000838    0.002515  0.012573   
cajun_creole  0.000000  0.000000  0.000000    0.000000  0.006849   
canadian      0.036176  0.000000  0.000000    0.000000  0.036176   

              apple_brandy   apricot  armagnac  artemisia  artichoke  ...  \
cuisine                                                               ...   
african           0.000000  0.069565    0.0000        0.0   0.000000  ...   
american          0.000623  0.011308    0.0001        0.0   0.006351  ...   
asian             0.000000  0.005029    0.0000        0.0   0.000000  ...   
cajun_creole      0.000000  0.000000    0.0000        0.0   0.000000  ...   
canadian          0.000000  0.002584    0.0000        0.0   0.001292  ...   

               whiskey  white_bread  white_wine  whole_grain_wheat_flour  \
cuisine                                                                    
african       0.000000     0.008696    0.043478                 0.008696   
american      0.002964     0.006874    0.030809                 0.014819   
asian         0.000838     0.001676    0.038558                 0.001676   
cajun_creole  0.000000     0.006849    0.082192                 0.000000   
canadian      0.002584     0.003876    0.029716                 0.020672   

                  wine      wood       yam     yeast    yogurt  zucchini  
cuisine                                                                   
african       0.017391  0.000000  0.008696  0.017391  0.000000  0.034783  
american      0.011009  0.000672  0.001445  0.068219  0.016912  0.018630  
asian         0.124895  0.000000  0.001676  0.004191  0.010897  0.011735  
cajun_creole  0.191781  0.000000  0.006849  0.034247  0.006849  0.000000  
canadian      0.003876  0.000000  0.001292  0.067183  0.019380  0.011628  

[5 rows x 383 columns]

Berikutnya, yuk kita cek 4 bahan paling populer untuk amsing2 masakan!

In [72]:
num_ingredients = 4 # jumlah bahan yang mau ditampilkan

# definisikan fungsi untuk short 4 bahan teratas
def print_top_ingredients(row):
    print(row.name.upper())
    row_sorted = row.sort_values(ascending=False)*100
    top_ingredients = list(row_sorted.index.values)[0:num_ingredients]
    row_sorted = list(row_sorted)[0:num_ingredients]

    for ind, ingredient in enumerate(top_ingredients):
        print("{} {:f}%".format(ingredient, row_sorted[ind]), end=' ')
    print("\n")

# apply function to cuisines dataframe
create_cuisines_profiles = cuisines.apply(print_top_ingredients, axis=1)

AFRICAN
onion 53.043478% olive_oil 52.173913% garlic 49.565217% cumin 42.608696% 

AMERICAN
butter 41.158157% egg 40.513076% wheat 39.840598% onion 29.332503% 

ASIAN
soy_sauce 49.622800% ginger 48.616932% garlic 47.946354% rice 41.324392% 

CAJUN_CREOLE
onion 69.863014% cayenne 56.164384% garlic 48.630137% butter 36.301370% 

CANADIAN
wheat 39.534884% butter 38.113695% egg 35.400517% onion 34.366925% 

CARIBBEAN
onion 51.366120% garlic 50.819672% vegetable_oil 31.147541% black_pepper 31.147541% 

CENTRAL_SOUTHAMERICAN
garlic 56.846473% onion 54.356846% cayenne 51.867220% tomato 41.493776% 

CHINESE
soy_sauce 68.552036% ginger 53.393665% garlic 52.941176% scallion 48.190045% 

EAST_ASIAN
garlic 55.205047% soy_sauce 50.368034% scallion 49.526814% cayenne 47.634069% 

EASTERN-EUROPE
wheat 53.191489% egg 52.340426% butter 48.085106% onion 45.106383% 

EASTERNEUROPEAN_RUSSIAN
butter 60.273973% egg 50.684932% wheat 49.315068% onion 38.356164% 

ENGLISH_SCOTTISH
butter 67.156863% wheat 62.25

## Terimakasih! 
<img src="http://multimodal.mipa.ugm.ac.id/wp-content/uploads/sites/371/2020/02/isnaalfi-youtube-01.png" style="width: 400px">